iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java

```diff
@@ -163,4 +165,32 @@ public static void updateSpec(Configuration configuration, Table table) {
   public static boolean isBucketed(Table table) {
     return table.spec().fields().stream().anyMatch(f -> f.transform().toString().startsWith("bucket["));
   }
+
+  /**
+   * Returns the snapshot ID which is immediately before (or exactly at) the timestamp provided in millis.
+   * If the timestamp provided is before the first snapshot of the table, we return an empty optional.
+   * If the timestamp provided is in the future compared to the latest snapshot, we return the latest snapshot ID.
+   *
+   * E.g.: if we have snapshots S1, S2, S3 committed at times T3, T6, T9 respectively (T0 = start of epoch), then:
+   * - from T0 to T2 -> returns empty
+   * - from T3 to T5 -> returns S1
+   * - from T6 to T8 -> returns S2
+   * - from T9 to T∞ -> returns S3
+   *
+   * @param table the table whose snapshot ID we are trying to find
+   * @param time the timestamp provided in milliseconds
+   * @return the snapshot ID corresponding to the time
+   */
+  public static Optional<Long> findSnapshotForTimestamp(Table table, long time) {
+    if (table.history().get(0).timestampMillis() > time) {
+      return Optional.empty();
+    }
+
+    for (Snapshot snapshot : table.snapshots()) {
```

Comment:
Are we certain that the table.snapshots() returns a list sorted by snapshot time?

Reply:
That's a good question. The snapshots come from `TableMetadata.snapshots()` which returns a `List<Snapshot>`.  The snapshots seem to be sorted by sequence number, which means it's also sorted by snapshot time millis:
https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/TableMetadata.java#L982-L990

Reply:
Actually this is only true for V2 tables. Let me debug into it a bit to see what's happening for V1

Reply:
Looks like the snapshots are ordered by commit time. 
Whenever there's a commit, we take the existing list of the snapshots [here](https://github.com/apache/iceberg/blob/9618147b6de8f8627052a205b86e45263394b0c2/core/src/main/java/org/apache/iceberg/TableMetadata.java#L817), and simply append the new snapshot to the end [here](https://github.com/apache/iceberg/blob/9618147b6de8f8627052a205b86e45263394b0c2/core/src/main/java/org/apache/iceberg/TableMetadata.java#L994). 
And since it's a List, the iteration order will be deterministic.

Reply:
Thanks for the explanation!